lift/src/main/scala/com/linkedin/lift/lib/PositionBiasUtils.scala

```diff
@@ -0,0 +1,191 @@
+    * @param data                dataset containing score, binary label {0, 1}, position
+    * @param maxImportanceWeight to control the variance of an importance sampling estimator
+    * @param targetPosition      target position for position bias estimation
+    * @param basePosition        base position for position bias estimation
+    * @return estimated position bias
+    */
+  def estimateAdjacentPositionBias(spark: SparkSession, data: Dataset[ScoreWithLabelAndPosition],
+    maxImportanceWeight: Double, targetPosition: Int, basePosition: Int): Double = {
+    import spark.implicits._
+
+    // estimating the density of the scores at basePosition
+    val kdTargetPosition = getDensity(extractScore(spark, data, targetPosition))
+
+    // estimating the density of the scores at basePosition
+    val kdPreviousPosition = getDensity(extractScore(spark, data, basePosition))
+
+    // extracting scores corresponding to positive labels at targetPosition
+    val targetPositionPositiveScoresArray = data
+      .filter($"label" === 1 and $"position" === targetPosition)
+      .map(r => r.score).collect
+
+    // estimating the positive label probability at targetPosition with importance sampling adjustment
+    val totalWeight = kdPreviousPosition.estimate(targetPositionPositiveScoresArray)
+      .zip(kdTargetPosition.estimate(targetPositionPositiveScoresArray))
+      .map(x => Math.min(x._1 / x._2, maxImportanceWeight)).sum
+
+    // estimating the positive label probability at basePosition
+    val basePositionPositiveScoresCount = data
+      .filter($"label" === 1 and $"position" === basePosition)
+      .count
+
+    return totalWeight / basePositionPositiveScoresCount
+  }
+
+  /**
+    * Estimating the position bias with respect to the top most position
+    * based on cumulative multiplications of estimated adjacent position biases.
+    * The adjacent position bias at a target position is estimated as the decay in the number of positive examples
+    * from the previous position to the target position after adjusting for
+    * the discrepancy in the score distributions with importance sampling (see above).
+    *
+    * @param spark                        spark session
+    * @param data                         dataset containing score, binary label {0, 1}, position
+    * @param maxImportanceWeight          to control the variance of an importance sampling estimator
+    *                                     for the adjacent position bias estimations
+    * @param positionBiasEstimationCutOff all adjacent position biases will be set to one beyond this cutoff
+    * @return position bias estimates for all positions
+    */
+  def estimatePositionBias(spark: SparkSession, data: Dataset[ScoreWithLabelAndPosition],
+    maxImportanceWeight: Double, positionBiasEstimationCutOff: Int): Seq[PositionBias] = {
+    import spark.implicits._
+    val positions = data.select($"position").distinct.map(r => r.getInt(0)).collect().toList.sorted
```

Comment:
Simplify as `data.map(_.position).distinct.toSeq.sorted`

Reply:
Thanks! Slightly modified your suggestion to
val positions = data.map(_.position).distinct.collect.toSeq.sorted